docs/decisions/0002-limiting-stream-length.rst

```diff
@@ -0,0 +1,68 @@
+0002 Limiting stream length
+###########################
+
+Status
+******
+
+**Provisional**
+
+Context
+*******
+
+The redis streams are an append-only data structure that stores entries
+forever. All the triggered events will be stored in the stream and eventually
+consume all the available memory. However storing these events forever is
+not required so we can set a limit to the number entries that can be stored,
+removing older entries whenever the limit is crossed.
+
+Decision
+********
+
+There are two options to cap these streams to a specific length.
+
+Running `XTRIM`_ command regularly by creating custom django command. It
+supports two strategies to evict older entries.
+
+.. _XTRIM: https://redis.io/commands/xtrim/
+
+
+* MAXLEN: Evicts entries as long as the stream's length exceeds the specified
+  threshold, where threshold is a positive integer.
+* MINID: Evicts entries with IDs lower than threshold, where threshold is a
+  stream ID.
+
+
+Another option which only provides single strategy i.e. ``MAXLEN`` but
+simpler in terms of implementation as well as maintenance is including maximum
+length of stream while adding items to the stream.
+
+.. code-block::
+
+   stream.add(data, id='*', maxlen=None, approximate=True)
+
+We can set maxlen value via django settings with a default value of 1000 (Bear
```

Comment:
Can we make a quick assessment of how much memory this would represent with our current uses? Our current use cases are pretty infrequent, but we have some really noisy ones that we're thinking about adding which may blow through 1k very quickly. If we can support 10k or more that would be ideal, and if not I'll need to be thinking about how we can change deployment to make it possible.

Reply:
@bmtcril I did a small test with `XBLOCK_DELETED` signal, published 40 events into a stream and checked its memory usage like so:

```
127.0.0.1:6379> MEMORY USAGE dev-xblock-deleted
(integer) 13227 <bytes>
```

The memory usage for a single event is around 350 bytes for this event but it might be a bit on the higher side for events with more data. So we can consider a value of 500 bytes per event.
So for 10K deleted events we need around 5 Mb of memory which doesn't seem a lot.

Reply:
@bmtcril Done, also changed the default value to 10K events. Let me know if I should revert to 1000.